DeviceAdapters/IntegratedLaserEngine/ILEWrapper/ILEWrapper.cpp

```diff
@@ -0,0 +1,367 @@
+///////////////////////////////////////////////////////////////////////////////
+// FILE:          ILEWrapper.cpp
+// PROJECT:       Micro-Manager
+// SUBSYSTEM:     DeviceAdapters
+//-----------------------------------------------------------------------------
+// DESCRIPTION:   IntegratedLaserEngine controller adapter
+//
+// Based off the AndorLaserCombiner adapter from Karl Hoover, UCSF
+//
+
+#ifdef WIN32
+#include <windows.h>
+#endif
+
+#include "ALC_REV.h"
+#include "ILEWrapper.h"
+#include "ALC_REVOject3Wrapper.h"
+#include "ALC_REV_ILEActiveBlankingManagementWrapper.h"
+#include "ALC_REV_ILEPowerManagementWrapper.h"
+#include "ALC_REV_ILE2Wrapper.h"
+#include "ALC_REV_ILE4Wrapper.h"
+#include "ILESDKLock.h"
+#include "../IntegratedLaserEngine.h"
+#include "../../../MMDevice/DeviceThreads.h"
+#include <sstream>
+#include <exception>
+
+///////////////////////////////////////////////////////////////////////////////
+// ILE Wrapper singleton handling
+///////////////////////////////////////////////////////////////////////////////
+
+/** This instance is shared between all ILE devices */
+static IILEWrapperInterface* ILEWrapper_s;
+MMThreadLock ILEWrapperLock_s;
+int CILEWrapper::NbInstances_s = 0;
+
+IILEWrapperInterface* LoadILEWrapper( CIntegratedLaserEngine* Caller )
+{
+  MMThreadGuard G( ILEWrapperLock_s );
+  if ( ILEWrapper_s == nullptr )
+  {
+    try
+    {
+      ILEWrapper_s = new CILEWrapper();
+    }
+    catch ( std::exception &e )
+    {
+      Caller->LogMMMessage( e.what() );
+      throw e;
+    }
+    CDeviceUtils::SleepMs( 100 );
+  }
+
+  ++CILEWrapper::NbInstances_s;
+  return ILEWrapper_s;
+}
+
+void UnloadILEWrapper()
+{
+  MMThreadGuard g( ILEWrapperLock_s );
+  --CILEWrapper::NbInstances_s;
+
+  if ( CILEWrapper::NbInstances_s == 0 )
+  {
+    delete ILEWrapper_s;
```

Comment:
I get the warning

```text
...\DeviceAdapters\IntegratedLaserEngine\ILEWrapper\ILEWrapper.cpp(65,5): warning C5205: delete of an abstract class 'IILEWrapperInterface' that has a non-virtual destructor results in undefined behavior
```

which I think can be fixed by simply adding `virtual ~IILEWrapperInterface() = default;` to `ILEWrapperInterface.h`.

Reply:
Good catch. Fixing this revealed a bug in the destructor of ```ILEWrapper```.